src/manifold/src/manifold.cpp

```diff
@@ -845,4 +848,141 @@ Manifold Manifold::Hull() const { return Hull(GetMesh().vertPos); }
+        verts.clear();
+      }
+    } while (vl.inc());
+  return output;
+}
+std::vector<Manifold> Manifold::Fracture(
+    const std::vector<glm::vec3>& pts,
+    const std::vector<float>& weights) const {
+  std::vector<glm::dvec3> highpVerts(pts.size());
+  std::vector<double> highpWeights(weights.size());
+  for (size_t i = 0; i < pts.size(); i++) {
+    highpVerts[i] = pts[i];
+    highpWeights[i] = weights[i];
+  }
+  return Fracture(highpVerts, highpWeights);
+}
+
+std::vector<Manifold> Manifold::ConvexDecomposition() const {
+  // Step 1. Get a list of all unique triangle faces with at least one reflex
+  // edge
+  std::unordered_set<int> uniqueReflexFaceSet;
+  const Impl& impl = *GetCsgLeafNode().GetImpl();
+  for (size_t i = 0; i < impl.halfedge_.size(); i++) {
+    Halfedge halfedge = impl.halfedge_[i];
+    int faceA = halfedge.face;
+    int faceB = impl.halfedge_[halfedge.pairedHalfedge].face;
+    glm::vec3 tangent =
+        glm::cross(impl.faceNormal_[faceA],
+                   impl.vertPos_[impl.halfedge_[i].endVert] -
+                       impl.vertPos_[impl.halfedge_[i].startVert]);
+    float tangentProjection = glm::dot(impl.faceNormal_[faceB], tangent);
+    //  If we've found a pair of reflex triangles, add them to the set
+    if (tangentProjection > 1e-6) {
+      uniqueReflexFaceSet.insert(faceA);
+      uniqueReflexFaceSet.insert(faceB);
+    }
+  }
+  std::vector<int> uniqueFaces;  // Copy to a vector for indexed access
+  uniqueFaces.insert(uniqueFaces.end(), uniqueReflexFaceSet.begin(),
+                     uniqueReflexFaceSet.end());
+
+  // Step 2. Calculate the Circumcircles (centers + radii) of these triangles
+  std::vector<glm::dvec3> circumcenters(uniqueFaces.size());
+  std::vector<double> circumradii(uniqueFaces.size());
+  Vec<glm::dvec3> joggledVerts(impl.vertPos_.size());
+  for (size_t i = 0; i < impl.vertPos_.size(); i++) {
+    joggledVerts[i] = impl.vertPos_[i];
+  }
+  for (size_t i = 0; i < uniqueFaces.size(); i++) {
+    glm::dvec4 circumcircle = impl.Circumcircle(joggledVerts, uniqueFaces[i]);
+    circumcenters[i] =
+        glm::dvec3(circumcircle.x, circumcircle.y, circumcircle.z);
+    circumradii[i] = circumcircle.w;
+  }
+
+  // Step 3. If any two circumcenters are identical, joggle one of the triangle
+  // vertices, TODO: and store in a hashmap
+  std::mt19937 mt(1337);
+  std::uniform_real_distribution<double> dist(0.0, 1.0);
+  double randOffset = 0.0;
+  for (size_t i = 0; i < circumcenters.size() - 1; i++) {
+    for (size_t j = i + 1; j < circumcenters.size(); j++) {
+      if (glm::distance(circumcenters[i], circumcenters[j]) < 1e-6) {
+        joggledVerts[impl.halfedge_[(uniqueFaces[i] * 3) + 0].startVert] +=
+            glm::dvec3(dist(mt), dist(mt), dist(mt)) * 1e-11;
+      }
+    }
+  }
+
+  // Step 4. Recalculate the circumcenters
+  for (size_t i = 0; i < uniqueFaces.size(); i++) {
+    glm::dvec4 circumcircle = impl.Circumcircle(joggledVerts, uniqueFaces[i]);
```

Comment:
Can we joggle the circumcircles instead of the verts? Overriding `vertPos_` in an `Impl` function is pretty strange.

Reply:
My ultimate consideration is to record which verts were joggled before computing the voronoi diagram, and then to unjoggle them afterward so that they occupy their original exact coordinates again...

I'm considering alternative strategies, like applying an invertible spatial distortion to all the input vertices, and then backwards to all of the output voronoi diagram vertices.... just have to find the right strategy that preserves precision...